cachix/src/Cachix/Client/PushQueue.hs

```diff
@@ -0,0 +1,73 @@
+{- Implements a queue with the following properties:
+
+- waits for queue to be fully pushed when exiting using ctrl-c (SIGINT)
+- allows stopping the producer
+- avoid duplicate pushing of the same store paths
+
+To safetly exit on demand, signal SIGINT.
+-}
+module Cachix.Client.PushQueue
+  ( pushQueue,
+    Queue,
+  )
+where
+
+import qualified Cachix.Client.Push as Push
+import Control.Concurrent.Async
+import Control.Concurrent.STM (TVar, modifyTVar', newTVarIO, readTVarIO)
+import qualified Control.Concurrent.STM.TBQueue as TBQueue
+import qualified Data.Set as S
+import Protolude
+import qualified System.Posix.Signals as Signals
+
+type StorePath = Text
+
+type Queue = TBQueue.TBQueue StorePath
+
+data WorkerState
+  = WorkerState
+      { inQueue :: Queue,
+        inProgress :: TVar (S.Set StorePath)
+      }
+
+worker :: Push.PushCache IO () -> WorkerState -> IO ()
+worker pushCache workerState = forever $ do
+  storePath <- atomically $ TBQueue.readTBQueue $ inQueue workerState
+  inprogress <- readTVarIO $ inProgress workerState
+  -- TODO: there's a small window between checking membership and inserting new path
```

Comment:
Your first `atomically` can take from the queue, update `inProgress` and return `Maybe StorePath`.
Then `for` the maybe, you do the work and at then end update `inProgress` in the second transaction.